Code/Tools/FBuild/Documentation/docs/functions/settings.html

```diff
@@ -51,6 +51,17 @@ <h1>Settings</h1>
   .AllowDBMigration_Experimental    // (optional) Migrate build state when bff changes to avoid rebuilds (default: false)
                                     // This enables what will eventually be standard behaviour
                                     // WARNING: This feature is experimental.
+  .WorkerListRefreshLimit           // (optional) When not using a static worker list,
+                                    //            max number of seconds to refresh the network share worker list.
+                                    //            default: 300 (5 minutes, in case workers are rebooting)
+                                    //            0  means no retries
+                                    //            -1 means infinite retries
+
+  .WorkerConnectionRetryLimit       // (optional) When a worker does not respond to a connection request,
+                                    //            max number of seconds to retry a worker connection.
+                                    //            default: 300 (5 minutes, in case workers are rebooting)
+                                    //            0  means no retries
+                                    //            -1 means infinite retries
```

Comment:
Does this mean "if a worker can't be connected to, try again after .WorkerConnectionRetryLimit amount of time" or does it mean "for the first .WorkerConnectionRetryLimit of a build, retry connections at some unspecified interval"?

Reply:
WorkerConnectionRetryLimit is not an interval. It is an upper limit similar to WorkerListRefreshLimit above.
So it represents the upper limit of time spent retrying a connection to a given worker.

It allows the client to stop trying a worker that is failing to connect.
Some use cases: a worker is rebooting, a worker app crashed, client or worker was firewalled, etc.
I default the WorkerConnectionRetryLimit to 5 minutes for the reboot case, since it is the largest known value of the use cases I could think of. The other use cases above are unknown time or may never recover & connect. Also, this makes the defaults the same for both the WorkerConnectionRetryLimit and the WorkerListRefreshLimit, which is easier for users to understand.

Reply:
Thanks for the clarification. I think we don't need this limit at all. I would expect that for the lifetime of the build, if the FBuildWorker hasn't reached the worker limit (default 15) it will continue to try to connect to available workers in the pool.

Reply:
I still think we need the WorkerConnectionRetryLimit, for these cases:

1. Worker crashed (or was killed by user) and brokerage file is still present. After some time, we want to exclude this worker from the 15 worker pool and not try to connect to it. Otherwise, we could build up a 15 worker pool list of machines that are all not connectable.

2. Worker is firewalled. Same reasons as above.

Thoughts?

Reply:
In the current version, workers that cannot be reached (for any reason) or actively refuse the connection, do not count in the worker limit, so neither of those cases should be a problem. We should make sure that behavior remains.

Reply:
Thanks. I see now that the failed connection workers do not apply to the 15 worker limit.
Thus, I concede that this connection limit is not needed for the case I mentioned.
But it may be useful as a performance optimization:

Say there are 16 machines that are in the brokerage path.
And 15 of those are not connectable (crashed, killed, firewalled, etc).
Then if we could exclude those 15 after some timeout time, then we could quickly connect to the 16th, 17th, etc machines in the next connection attempt.

Granted, this would mean we would never retry the 15 machines we timed out on. This could be bad if they at some point were restored: user relaunched the worker, unfirewalled the machine, etc.)

I guess I am worried that without a timeout we would be slow to get to the 16th, 17th, etc machines, since the current algorithm randomly chooses a machine from the brokerage path.
Will the random number generator eventually get to the 16th and 17th machines?
How long will it take? Perhaps a priority queue would be a better improvement over a timeout here, to solve this performance issue?

Thoughts?